lib/dynflow/action/with_sub_plans.rb

```diff
@@ -2,33 +2,58 @@ module Dynflow
   module Action::WithSubPlans
     include Dynflow::Action::Cancellable
 
+    ExecutionTicket = Algebrick.type do
+      fields! :count => type { variants NilClass, Integer }
+    end
+
+    TimeTicket = Algebrick.type do
+      fields! :interval => type { variants NilClass, Integer },
+              :execution_ticket  => type { variants ExecutionTicket, NilClass }
+    end
+
     SubPlanFinished = Algebrick.type do
       fields! :execution_plan_id => String,
-              :success           => type { variants TrueClass, FalseClass }
+              :success           => type { variants TrueClass, FalseClass },
+              :execution_ticket  => type { variants ExecutionTicket, NilClass }
     end
 
     def run(event = nil)
       match event,
             (on nil do
-               if output[:total_count]
-                 resume
-               else
-                 initiate
-               end
+               start
+             end),
+            (on ExecutionTicket.(~any) do |count|
+               start count
              end),
             (on SubPlanFinished do
-              mark_as_done(event.execution_plan_id, event.success)
-              try_to_finish or suspend
+               mark_as_done(event.execution_plan_id, event.success)
+               try_to_finish or wake_n(event.execution_ticket.nil? ? 0 : event.execution_ticket.count) or suspend
```

Comment:
the method could be made shorter by extracting the bodies of contiions into separate methods. Especially this line would help being extracted into multiple lines